src/GraphBLAS-sharp/Abstracts.fs

```diff
@@ -12,56 +14,48 @@ type Matrix<'a when 'a : struct and 'a : equality>() =
-    abstract EWiseAdd: Matrix<'a> -> Mask2D option -> Semiring<'a> -> Matrix<'a>
-    abstract EWiseMult: Matrix<'a> -> Mask2D option -> Semiring<'a> -> Matrix<'a>
+    abstract Fill: Mask2D option -> Scalar<'a> with set
+    abstract Fill: Mask1D option * int -> Scalar<'a> with set
+    abstract Fill: int * Mask1D option -> Scalar<'a> with set
+
+    abstract Mxm: Matrix<'b> -> Mask2D option -> Semiring<'a, 'b, 'c> -> Matrix<'c>
+    abstract Mxv: Vector<'b> -> Mask1D option -> Semiring<'a, 'b, 'c> -> Vector<'c>
+    abstract EWiseAdd: Matrix<'a> -> Mask2D option -> BinaryOp<'a, 'b, 'c> -> Matrix<'a>
+    abstract EWiseMult: Matrix<'a> -> Mask2D option -> BinaryOp<'a, 'b, 'c> -> Matrix<'a>
     abstract Apply: Mask1D option -> UnaryOp<'a, 'b> -> Matrix<'b>
-    abstract ReduceIn: Mask1D option -> Monoid<'a> -> Vector<'a>
-    abstract ReduceOut: Mask1D option -> Monoid<'a> -> Vector<'a>
+    abstract ReduceIn: Mask1D option -> BinaryOp<'a, 'b, 'c> -> Vector<'a>
+    abstract ReduceOut: Mask1D option -> BinaryOp<'a, 'b, 'c> -> Vector<'a>
+    abstract Reduce: Monoid<'a> -> Scalar<'a>
     abstract T: Matrix<'a>
 
-    abstract EWiseAddInplace: Matrix<'a> -> Mask2D option -> Semiring<'a> -> unit
-    abstract EWiseMultInplace: Matrix<'a> -> Mask2D option -> Semiring<'a> -> unit
-    abstract ApplyInplace: Mask2D option -> UnaryOp<'a, 'b> -> unit
-
     static member inline (+) (x: Matrix<'a>, y: Matrix<'a>) = x.EWiseAdd y
     static member inline (*) (x: Matrix<'a>, y: Matrix<'a>) = x.EWiseMult y
-    static member inline (+.*) (x: Matrix<'a>, y: Matrix<'a>) = x.Mxm y
-    static member inline (+.*) (x: Matrix<'a>, y: Vector<'a>) = x.Mxv y
-    static member inline (.+) (x: Matrix<'a>, y: Matrix<'a>) = x.EWiseAddInplace y
-    static member inline (.*) (x: Matrix<'a>, y: Matrix<'a>) = x.EWiseMultInplace y
+    static member inline (@.) (x: Matrix<'a>, y: Matrix<'b>) = x.Mxm y
+    static member inline (@.) (x: Matrix<'a>, y: Vector<'b>) = x.Mxv y
 
-and [<AbstractClass>] Vector<'a when 'a : struct and 'a : equality>() =
+and [<AbstractClass>] Vector<'a when 'a : struct and 'a : equality>(length: int) =
     abstract Length: int
+    default this.Length = length
+
     abstract AsArray: 'a[]
+    abstract Clear: unit -> unit
 
     abstract Mask: Mask1D option
     abstract Complemented: Mask1D option
 
     abstract Item: Mask1D option -> Vector<'a> with get, set
     abstract Item: int -> Scalar<'a> with get, set
-    abstract Item: Mask1D option -> Scalar<'a> with set
+    abstract Fill: Mask1D option -> Scalar<'a> with set
 
-    abstract Vxm: Matrix<'a> -> Mask1D option -> Semiring<'a> -> Vector<'a>
-    abstract EWiseAdd: Vector<'a> -> Mask1D option -> Semiring<'a> -> Vector<'a>
-    abstract EWiseMult: Vector<'a> -> Mask1D option -> Semiring<'a> -> Vector<'a>
+    abstract Vxm: Matrix<'b> -> Mask1D option -> Semiring<'a, 'b, 'c> -> Vector<'c>
+    abstract EWiseAdd: Vector<'a> -> Mask1D option -> BinaryOp<'a, 'b, 'c> -> Vector<'a>
```

Comment:
Looks like we should use monoid instead of binaryOp, because for sparse operations we need to know an ID. Thus, it may be useful to define semiring as an extension of monoids.